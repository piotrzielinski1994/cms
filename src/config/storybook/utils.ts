import { ReactNode } from 'react';

const DEFAULT_VALUE = '__default__' as const;
const THUMBNAIL_ID = 'thumbnail';

const storyToUrlPath = (title: string) => {
  const parts = title.split('/');
  const story = parts[parts.length - 1];
  const titlePart = title.toLowerCase().replace(/\//g, '-').replace(/\s+/g, '-');
  const storyPart = story
    .replace(/([a-z0-9])([A-Z])/g, '$1-$2')
    .replace(/([a-zA-Z])(\d+)/g, '$1-$2')
    .toLowerCase();
  return `${titlePart}--${storyPart}`;
};

const getFallback = <T extends ReactNode>(value: T, translation: string): T => {
  return value === DEFAULT_VALUE ? (translation as T) : value;
};

export { DEFAULT_VALUE, getFallback, storyToUrlPath, THUMBNAIL_ID };
