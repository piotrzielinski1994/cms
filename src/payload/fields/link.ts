import deepMerge from '@/_old/utilities/deepMerge';
import { AdminTranslations } from '@/config/locales.config';
import type { Field } from 'payload';
export type LinkAppearances = 'default' | 'outline';

// Types ====================================

type LinkType = (options?: {
  appearances?: LinkAppearances[] | false;
  disableLabel?: boolean;
  overrides?: Record<string, unknown>;
  depth?: number;
}) => Field;

// Variables ====================================

const appearanceOptions: Record<LinkAppearances, { label: string; value: string }> = {
  default: {
    label: 'Default',
    value: 'default',
  },
  outline: {
    label: 'Outline',
    value: 'outline',
  },
};

const link: LinkType = ({ depth = 0, appearances, disableLabel = false, overrides = {} } = {}) => {
  const linkResult: Field = {
    name: 'link',
    type: 'group',
    label: ({ t }: { t: AdminTranslations }) => t('fields:type'),
    admin: {
      hideGutter: true,
    },
    fields: [
      {
        type: 'row',
        fields: [
          {
            name: 'type',
            type: 'radio',
            label: ({ t }: { t: AdminTranslations }) => t('fields:navItem'),
            admin: {
              layout: 'horizontal',
              width: '50%',
            },
            defaultValue: 'reference',
            options: [
              {
                label: ({ t }: { t: AdminTranslations }) => t('fields:internalLink'),
                value: 'reference',
              },
              {
                label: ({ t }: { t: AdminTranslations }) => t('fields:customURL'),
                value: 'custom',
              },
            ],
          },
          {
            name: 'newTab',
            type: 'checkbox',
            admin: {
              style: {
                alignSelf: 'flex-end',
              },
              width: '50%',
            },
            label: ({ t }: { t: AdminTranslations }) => t('common:openInNewTab'),
          },
        ],
      },
    ],
  };

  const linkTypes: Field[] = [
    {
      name: 'reference',
      type: 'relationship',
      admin: {
        condition: (_, siblingData) => siblingData?.type === 'reference',
      },
      label: ({ t }: { t: AdminTranslations }) => t('fields:documentToLinkTo'),
      relationTo: ['pages'],
      required: true,
    },
    {
      name: 'url',
      type: 'text',
      admin: {
        condition: (_, siblingData) => siblingData?.type === 'custom',
      },
      label: ({ t }: { t: AdminTranslations }) => t('fields:customURL'),
      required: true,
    },
  ];

  if (!disableLabel) {
    linkTypes.map((linkType) => ({
      ...linkType,
      admin: {
        ...linkType.admin,
        width: '50%',
      },
    }));

    linkResult.fields.push({
      type: 'row',
      fields: [
        ...linkTypes,
        {
          name: 'label',
          type: 'text',
          admin: {
            width: '50%',
          },
          label: ({ t }: { t: AdminTranslations }) => t('fields:label'),
          required: true,
          localized: true,
        },
      ],
    });
  } else {
    linkResult.fields = [...linkResult.fields, ...linkTypes];
  }

  if (appearances !== false) {
    let appearanceOptionsToUse = [appearanceOptions.default, appearanceOptions.outline];

    if (appearances) {
      appearanceOptionsToUse = appearances.map((appearance) => appearanceOptions[appearance]);
    }

    linkResult.fields.push({
      name: 'appearance',
      type: 'select',
      admin: {
        description: 'Choose how the link should be rendered.',
      },
      defaultValue: 'default',
      options: appearanceOptionsToUse,
    });
  }

  if (depth > 0) {
    linkResult.fields.push({
      name: 'navItems',
      type: 'array',
      label: ({ t }: { t: AdminTranslations }) => t('fields:submenu'),
      fields: [
        link({
          appearances: false,
          depth: depth - 1,
        }),
      ],
      maxRows: 6,
      admin: {
        style: {
          paddingLeft: 200,
        },
        initCollapsed: true,
        components: {
          RowLabel: '@/payload/components/row-label#RowLabel',
        },
      },
    });
  }

  return deepMerge(linkResult, overrides);
};

export { appearanceOptions, link };
