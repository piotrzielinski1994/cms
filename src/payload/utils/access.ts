import { User } from '@/payload.types';
import { Access, FieldAccess } from 'payload';

const anyone: Access = () => true;

const authenticated: Access = ({ req: { user } }) => {
  return Boolean(user);
};

const admin: Access = ({ req: { user } }) => {
  return Boolean(user?.roles?.includes('admin'));
};

const adminOrSelf: Access = ({ req: { user } }) => {
  if (!user) return false;
  if (user.roles?.includes('admin')) return true;
  return { id: { equals: user.id } };
};

const adminFieldLevel: FieldAccess<{ id: string }, User> = ({ req: { user } }) => {
  return Boolean(user?.roles?.includes('admin'));
};

const authenticatedOrPublished: Access = ({ req: { user } }) => {
  if (user) return true;
  return { _status: { equals: 'published' } };
};

export { admin, adminFieldLevel, adminOrSelf, anyone, authenticated, authenticatedOrPublished };
