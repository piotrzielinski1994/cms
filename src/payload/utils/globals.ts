import type { Config } from '@/payload.types';
import configPromise from '@/payload/payload.config';
import { Locale } from 'next-intl';
import { unstable_cache } from 'next/cache';
import { getPayload } from 'payload';

const getGlobal = async (slug: keyof Config['globals'], locale: Locale = 'en', depth = 1) => {
  const payload = await getPayload({ config: configPromise });
  return payload.findGlobal({ slug, depth, locale });
};

const getCachedGlobal = (slug: keyof Config['globals'], locale: Locale, depth = 1) => {
  const cacheConfig = { tags: [`global__${locale}_${slug}`] };
  return unstable_cache(() => getGlobal(slug, locale, depth), [slug, locale], cacheConfig);
};

export { getCachedGlobal };
