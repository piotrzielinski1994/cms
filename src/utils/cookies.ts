import { setCookie } from 'typescript-cookie';

const setPermament = (name: string, value: string | number | boolean | object) => {
  setCookie(name, value.toString(), { expires: 365 * 10, path: '/' });
};

const cookies = { setPermament };

export default cookies;
