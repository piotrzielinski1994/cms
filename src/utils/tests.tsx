import { Providers } from '@/providers';
import { ComponentProps, ReactNode } from 'react';

const resolveServerComponent = async <T extends (props: unknown) => Promise<ReactNode>>(
  Component: T,
  props: Parameters<T>[0],
) => {
  const ComponentResolved = await Component(props);
  return () => ComponentResolved;
};

const withProviders = (
  children: ReactNode,
  props: Partial<Omit<ComponentProps<typeof Providers>, 'children'>> = {},
) => {
  const overwrittenProps: ComponentProps<typeof Providers> = {
    locale: 'en',
    initialTheme: 'light',
    initialFontScale: 'base',
    initialCookiesConsent: true,
    ...props,
    children,
  };
  return <Providers {...overwrittenProps} />;
};

export { resolveServerComponent, withProviders };
