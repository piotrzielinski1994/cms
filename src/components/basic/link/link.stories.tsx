import { DEFAULT_VALUE, getFallback } from '@/config/storybook/utils';
import type { Meta, StoryObj } from '@storybook/react';
import { useTranslations } from 'next-intl';
import { type ComponentProps } from 'react';
import { Link as LinkComponent } from './link';

type Args = ComponentProps<typeof LinkComponent>;

const meta = {
  component: LinkComponent,
  title: 'Basic/Link',
  argTypes: {
    href: { control: 'text' },
    children: { control: 'text' },
  },
  args: {
    href: '/',
    children: DEFAULT_VALUE,
  },
} satisfies Meta<Args>;

const Render = ({ children, ...args }: Args) => {
  const t = useTranslations('common');
  return <LinkComponent {...args}>{getFallback(children, t('link'))}</LinkComponent>;
};

const Link: StoryObj<typeof LinkComponent> = { render: Render };

export { Link };
export default meta;
