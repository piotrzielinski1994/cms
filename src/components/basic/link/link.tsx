import { Link as LocalizedLink } from '@/config/next.routing.config';
import { cn } from '@/utils/tailwind';
import { ComponentProps } from 'react';

const Link = ({ className, ...rest }: ComponentProps<typeof LocalizedLink>) => {
  return (
    <LocalizedLink
      {...rest}
      className={cn('text-primary underline', 'focus-visible:tw-cms-outline', className)}
    />
  );
};

export { Link };
