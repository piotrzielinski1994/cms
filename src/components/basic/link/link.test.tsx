import { withProviders } from '@/utils/tests';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { ComponentProps } from 'react';
import { describe, expect, it, vi } from 'vitest';
import { axe } from 'vitest-axe';
import { Link } from './link';

describe('Link', () => {
  const defaultProps = {
    href: '/test',
    children: 'Link',
  } satisfies ComponentProps<typeof Link>;

  beforeEach(() => {
    vi.resetAllMocks();
  });

  it('should have no accessibility violations', async () => {
    const { container } = render(withProviders(<Link {...defaultProps} />));
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should match the snapshot', () => {
    const { container } = render(withProviders(<Link {...defaultProps} />));
    expect(container).toMatchSnapshot();
  });

  it('calls onClick handler', async () => {
    const onClick = vi.fn();
    const { getByRole } = render(withProviders(<Link {...defaultProps} onClick={onClick} />));

    await userEvent.click(getByRole('link', { name: defaultProps.children }));

    expect(onClick).toHaveBeenCalled();
  });
});
