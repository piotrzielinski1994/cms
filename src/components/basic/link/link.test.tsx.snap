// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Link > should match the snapshot 1`] = `
<div>
  <a
    class="text-primary underline focus-visible:tw-cms-outline"
    href="/test"
  >
    Link
  </a>
</div>
`;
